In [1]:
import pandas as pd
import numpy as np

In [2]:
df_payment = pd.read_csv('/content/payment_report.csv')
df_transactions= pd.read_csv('/content/transactions.csv')
df_product = pd.read_csv('/content/product.csv')

**PART 1: EDA**

**1.1 EDA df payment_enriched**

In [3]:
#Create df_payment_enriched (Merge payment + product)
df_payment_enriched = pd.merge(df_payment,df_product,how = 'left',on = 'product_id')
df_payment_enriched

,report_month,payment_group,product_id,source_id,volume,category,team_own
0,2023-01,payment,12,45,624110375,PXXXXXT,ASD
1,2023-01,payment,17,45,335715113,PXXXXXB,ASD
2,2023-01,payment,18,45,737784466,PXXXXXB,ASD
3,2023-01,payment,19,45,120963069,PXXXXXM2,ASD
4,2023-01,payment,20,45,319653158,PXXXXXB,ASD
...,...,...,...,...,...,...,...
914,2023-04,payment,15067,45,1504000,PXXXXXR,ASL
915,2023-04,refund,1976,37,3542271587,NaN,NaN
916,2023-04,refund,1976,38,13831708189,NaN,NaN
917,2023-04,refund,1976,39,1905435543,NaN,NaN


In [4]:
#Check data type
df_payment_enriched.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 919 entries, 0 to 918
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   report_month   919 non-null    object
 1   payment_group  919 non-null    object
 2   product_id     919 non-null    int64 
 3   source_id      919 non-null    int64 
 4   volume         919 non-null    int64 
 5   category       897 non-null    object
 6   team_own       897 non-null    object
dtypes: int64(3), object(4)
memory usage: 50.4+ KB


In [5]:
#Convert report_month → datetime
df_payment_enriched['report_month'] = pd.to_datetime(df_payment_enriched['report_month'], format='%Y-%m')

In [6]:
#Check missing data
df_payment_enriched.isnull().sum()

,0
report_month,0
payment_group,0
product_id,0
source_id,0
volume,0
category,22
team_own,22


In [7]:
#Check duplicate
df_payment_enriched.duplicated().sum()

np.int64(0)

In [8]:
df_payment_enriched['volume'].describe()

,volume
count,9.190000e+02
mean,1.978574e+08
std,8.367595e+08
min,5.500000e+03
25%,1.250000e+06
50%,7.982015e+06
75%,5.447599e+07
max,1.383171e+10


In [9]:
df_payment_enriched.nunique()

,0
report_month,4
payment_group,2
product_id,311
source_id,4
volume,896
category,15
team_own,3


In [10]:
df_payment_enriched['report_month'].value_counts()

,count
report_month,
2023-04-01,281
2023-03-01,224
2023-02-01,218
2023-01-01,196


In [11]:
df_payment_enriched['payment_group'].value_counts()

,count
payment_group,
payment,902
refund,17


In [12]:
df_payment_enriched['category'].value_counts()

,count
category,
PXXXXXR,246
PXXXXXF,206
PXXXXXB,167
PXXXXXE,59
PXXXXXT2,44
PXXXXXM2,31
PXXXXXD1,28
PXXXXXT,27
PXXXXXS,27


In [13]:
df_payment_enriched['team_own'].value_counts()

,count
team_own,
ASL,523
ASD,340
APS,34




*   Size: 919 rows and 7 columns.
*   Data Types:
      *   Integer (int64): product_id, source_id, volume.
      *   Object (String): report_month, payment_group, category, team_own.
*   Missing Values: category and team_own have 22 null values (representing unclassified products) -> These are kept to maintain data integrity.
*   Duplicates: None.
*   Data Quality: No anomalies detected.

-> Action Required: Convert report_month from object to datet



** 1.2 EDA df_transactions**

In [14]:
df_transactions

,transaction_id,merchant_id,volume,transType,transStatus,sender_id,receiver_id,extra_info,timeStamp
0,3002692434,5,100000,24,1,10199794.0,199794.0,NaN,1682932054455
1,3002692437,305,20000,2,1,14022211.0,14022211.0,NaN,1682932054912
2,3001960110,7255,48605,22,1,NaN,10530940.0,NaN,1682932055000
3,3002680710,2270,1500000,2,1,10059206.0,59206.0,NaN,1682932055622
4,3002680713,2275,90000,2,1,10004711.0,4711.0,NaN,1682932056197
...,...,...,...,...,...,...,...,...,...
1323997,3003723030,305,20000,2,1,24524311.0,NaN,NaN,1683035672634
1323998,3003723033,2270,100000,2,1,10277242.0,277242.0,NaN,1683035672876
1323999,3003723036,2270,100000,2,1,10144599.0,144599.0,NaN,1683035672892
1324000,3003723039,5,400,22,1,10028007.0,21013762.0,NaN,1683035672896


In [15]:
df_transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1324002 entries, 0 to 1324001
Data columns (total 9 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   transaction_id  1324002 non-null  int64  
 1   merchant_id     1324002 non-null  int64  
 2   volume          1324002 non-null  int64  
 3   transType       1324002 non-null  int64  
 4   transStatus     1324002 non-null  int64  
 5   sender_id       1274943 non-null  float64
 6   receiver_id     1159207 non-null  float64
 7   extra_info      6095 non-null     object 
 8   timeStamp       1324002 non-null  int64  
dtypes: float64(2), int64(6), object(1)
memory usage: 90.9+ MB


In [16]:
#Convert timeStamp → datetime
df_transactions['timeStamp'] = pd.to_datetime(df_transactions['timeStamp'], unit='ms')

In [17]:
#Check missing data
df_transactions.isnull().sum()

,0
transaction_id,0
merchant_id,0
volume,0
transType,0
transStatus,0
sender_id,49059
receiver_id,164795
extra_info,1317907
timeStamp,0


In [18]:
#Check duplicate
df_transactions.duplicated().sum()

np.int64(28)

In [19]:
#Remove duplicate
df_transactions = df_transactions.drop_duplicates()

In [20]:
df_transactions['volume'].describe()

,volume
count,1.323974e+06
mean,2.388093e+05
std,9.681107e+05
min,1.000000e+00
25%,1.000000e+04
50%,3.000000e+04
75%,1.000000e+05
max,7.869148e+07


In [21]:
df_payment_enriched['report_month'].min(), df_payment_enriched['report_month'].max()

(Timestamp('2023-01-01 00:00:00'), Timestamp('2023-04-01 00:00:00'))

In [22]:
df_transactions.nunique()

,0
transaction_id,1323974
merchant_id,646
volume,49488
transType,7
transStatus,35
sender_id,173557
receiver_id,200291
extra_info,2037
timeStamp,1316590


In [23]:
df_transactions['transStatus'].value_counts()

,count
transStatus,
1,1248752
-217,67088
-358,2139
-351,1160
-357,1050
-68,961
-332,675
-333,313
-994,294


In [24]:
df_transactions['transType'].value_counts()

,count
transType,
2,760767
8,342549
22,202376
30,14585
24,3677
14,10
12,10


Size: 1,324,002 rows and 9 columns.

Data Types:

*   Integer (int64): transaction_id, merchant_id, volume, transType, transStatus, timeStamp.
*   Float (float64): sender_id, receiver_id.
*   Object (String): extra_info.
  -> Convert timeStamp from Unix milliseconds (int64) to datetime format.

Missing Values:
*   sender_id (~49K) contain nulls
*   receiver_id (~164K) contain nulls
*   extra_info (~1.3M) contain nulls

  -> These are business-valid (representing top-ups, withdrawals, or optional data) and are kept to maintain data integrity.

Duplicates: 28 duplicate rows identified and removed.

Data Quality:


*   transStatus: Categorized into Success (1) and Failure/Canceled (< 0).
*   volume: Ranges from 1 to 78.7M VND.

*   transType: Includes types 2, 8 (valid) and 24, 22, 30, 14, 12 (invalid but retained for full audit trail).


->
Final dataset size after cleaning: 1,323,974 rows and 9 columns.

**PART 2: Data Wrangling**

1.Top 3 product_ids with the highest volume

In [25]:
top3_products = (
    df_payment_enriched
    .groupby('product_id', as_index=False)['volume']  # Nhóm theo product
    .sum()                                              # Tổng volume
    .sort_values('volume', ascending=False)             # Sắp xếp giảm dần
    .head(3)                                            # Lấy top 3
    .reset_index(drop=True)
)
top3_products.index += 1
top3_products

,product_id,volume
1,1976,61797583647
2,429,14667676567
3,372,13713658515


In [26]:
total_vol = df_payment_enriched['volume'].sum()
top3_products['contribution_%'] = (
    top3_products['volume'] / total_vol * 100
).round(2)
top3_products

,product_id,volume,contribution_%
1,1976,61797583647,33.99
2,429,14667676567,8.07
3,372,13713658515,7.54


2.Given that 1 product_id is only owed by 1 team, are there any abnormal products against this rule?

In [27]:
teams_per_product = (
    df_product
    .groupby('product_id')['team_own']
    .nunique()                     # Đếm số team unique theo product
    .reset_index()
    .rename(columns={'team_own': 'num_teams'})
)

abnormal = teams_per_product[teams_per_product['num_teams'] > 1]
abnormal

,product_id,num_teams


3.Find the team has had the lowest performance (lowest volume) since Q2.2023. Find the category that contributes the least to that team

In [28]:
q2_start = pd.Timestamp('2023-04-01')
df_q2 = df_payment_enriched[df_payment_enriched['report_month'] >= q2_start].copy()

In [29]:
team_volume = (
    df_q2
    .dropna(subset=['team_own'])
    .groupby('team_own')['volume']
    .sum()
    .sort_values()
    .reset_index()
)
team_volume.columns = ['team_own', 'total_volume']
team_volume

,team_own,total_volume
0,APS,51141753
1,ASL,11284361730
2,ASD,31090428473


In [30]:
worst_team = team_volume.iloc[0]['team_own']
worst_team

'APS'

In [31]:
category_volume = (
    df_q2[df_q2['team_own'] == worst_team]
    .groupby('category')['volume']
    .sum()
    .sort_values()
    .reset_index()
)

category_volume.columns = ['category', 'total_volume']
total_team_vol = category_volume['total_volume'].sum()
category_volume['contribution_%'] = (
    category_volume['total_volume'] / total_team_vol * 100
).round(2)

category_volume

,category,total_volume,contribution_%
0,PXXXXXE,25232438,49.34
1,PXXXXXS,25909315,50.66


4.Find the contribution of source_ids of refund transactions (payment_group = ‘refund’), what is the source_id with the highest contribution?

In [49]:
df_refund = df_payment_enriched[
    df_payment_enriched['payment_group'] == 'refund'
].copy()
source_contribution = (
    df_refund
    .groupby('source_id')['volume']
    .sum()
    .reset_index()
    .sort_values('volume', ascending=False)
)

total_refund_vol = source_contribution['volume'].sum()
source_contribution['contribution_%'] = (
    source_contribution['volume'] / total_refund_vol * 100
).round(2)
source_contribution

,source_id,volume,contribution_%
1,38,36527454759,59.11
2,39,16119059662,26.08
0,37,9151069226,14.81


In [33]:
source_contribution.head(1)

,source_id,volume,contribution_%
1,38,36527454759,59.11


5.Define type of transactions (‘transaction_type’) for each row, given:
- transType = 2 & merchant_id = 1205: Bank Transfer Transaction
- transType = 2 & merchant_id = 2260: Withdraw Money Transaction
- transType = 2 & merchant_id = 2270: Top Up Money Transaction
- transType = 2 & others merchant_id: Payment Transaction
- transType = 8, merchant_id = 2250: Transfer Money Transaction
- transType = 8 & others merchant_id: Split Bill Transaction
- Remained cases are invalid transactions


In [34]:
df_transactions

,transaction_id,merchant_id,volume,transType,transStatus,sender_id,receiver_id,extra_info,timeStamp
0,3002692434,5,100000,24,1,10199794.0,199794.0,NaN,2023-05-01 09:07:34.455
1,3002692437,305,20000,2,1,14022211.0,14022211.0,NaN,2023-05-01 09:07:34.912
2,3001960110,7255,48605,22,1,NaN,10530940.0,NaN,2023-05-01 09:07:35.000
3,3002680710,2270,1500000,2,1,10059206.0,59206.0,NaN,2023-05-01 09:07:35.622
4,3002680713,2275,90000,2,1,10004711.0,4711.0,NaN,2023-05-01 09:07:36.197
...,...,...,...,...,...,...,...,...,...
1323997,3003723030,305,20000,2,1,24524311.0,NaN,NaN,2023-05-02 13:54:32.634
1323998,3003723033,2270,100000,2,1,10277242.0,277242.0,NaN,2023-05-02 13:54:32.876
1323999,3003723036,2270,100000,2,1,10144599.0,144599.0,NaN,2023-05-02 13:54:32.892
1324000,3003723039,5,400,22,1,10028007.0,21013762.0,NaN,2023-05-02 13:54:32.896


In [35]:
conditions = [
    (df_transactions['transType'] == 2) & (df_transactions['merchant_id'] == 1205),
    (df_transactions['transType'] == 2) & (df_transactions['merchant_id'] == 2260),
    (df_transactions['transType'] == 2) & (df_transactions['merchant_id'] == 2270),
    (df_transactions['transType'] == 2),
    (df_transactions['transType'] == 8) & (df_transactions['merchant_id'] == 2250),
    (df_transactions['transType'] == 8)
]

choices = [
    'Bank Transfer Transaction',
    'Withdraw Money Transaction',
    'Top Up Money Transaction',
    'Payment Transaction',
    'Transfer Money Transaction',
    'Split Bill Transaction'
]

df_transactions['transaction_type'] = np.select(conditions, choices, default='Invalid Transaction')

df_transactions

/tmp/ipykernel_3850/3872036258.py:19: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_transactions['transaction_type'] = np.select(conditions, choices, default='Invalid Transaction')


,transaction_id,merchant_id,volume,transType,transStatus,sender_id,receiver_id,extra_info,timeStamp,transaction_type
0,3002692434,5,100000,24,1,10199794.0,199794.0,NaN,2023-05-01 09:07:34.455,Invalid Transaction
1,3002692437,305,20000,2,1,14022211.0,14022211.0,NaN,2023-05-01 09:07:34.912,Payment Transaction
2,3001960110,7255,48605,22,1,NaN,10530940.0,NaN,2023-05-01 09:07:35.000,Invalid Transaction
3,3002680710,2270,1500000,2,1,10059206.0,59206.0,NaN,2023-05-01 09:07:35.622,Top Up Money Transaction
4,3002680713,2275,90000,2,1,10004711.0,4711.0,NaN,2023-05-01 09:07:36.197,Payment Transaction
...,...,...,...,...,...,...,...,...,...,...
1323997,3003723030,305,20000,2,1,24524311.0,NaN,NaN,2023-05-02 13:54:32.634,Payment Transaction
1323998,3003723033,2270,100000,2,1,10277242.0,277242.0,NaN,2023-05-02 13:54:32.876,Top Up Money Transaction
1323999,3003723036,2270,100000,2,1,10144599.0,144599.0,NaN,2023-05-02 13:54:32.892,Top Up Money Transaction
1324000,3003723039,5,400,22,1,10028007.0,21013762.0,NaN,2023-05-02 13:54:32.896,Invalid Transaction


In [36]:
df_transactions['transaction_type'].value_counts()

,count
transaction_type,
Payment Transaction,398665
Transfer Money Transaction,341173
Top Up Money Transaction,290498
Invalid Transaction,220658
Bank Transfer Transaction,37879
Withdraw Money Transaction,33725
Split Bill Transaction,1376


6.Of each transaction type (excluding invalid transactions): find the number of transactions, volume, senders and receivers.

In [37]:
df_valid = df_transactions[df_transactions['transaction_type'] != 'Invalid Transaction'].copy()

summary = df_valid.groupby('transaction_type').agg(
    num_transactions=('transaction_id', 'count'),
    total_volume=('volume', 'sum'),
    num_senders=('sender_id', 'nunique'),
    num_receivers=('receiver_id', 'nunique')
).reset_index()

# Format
for col in summary.select_dtypes(include='number').columns:
    summary[col] = summary[col].apply(lambda x: "{:,.0f}".format(x))

summary


,transaction_type,num_transactions,total_volume,num_senders,num_receivers
0,Bank Transfer Transaction,"37,879","50,605,806,190","23,156","9,271"
1,Payment Transaction,"398,665","71,850,608,441","139,583","113,298"
2,Split Bill Transaction,"1,376","4,901,464","1,323",572
3,Top Up Money Transaction,"290,498","108,605,618,829","110,409","110,409"
4,Transfer Money Transaction,"341,173","37,032,880,492","39,021","34,585"
5,Withdraw Money Transaction,"33,725","23,418,181,420","24,814","24,814"
